File: ArchitectureModel/src/main/java/com/kailin/architecture_model/encrypt/CipherUtil.java
package com.kailin.architecture_model.encrypt;

import android.util.Base64;

import java.security.Key;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

abstract class CipherUtil {

    protected final String ALGORITHM_FORMAT = "%s/%s/%s";

    protected abstract String getKeyString();

    protected abstract IvParameterSpec getIvParameterSpec();

    protected abstract String getAlgorithm();

    protected abstract String getTransformation();

    public String encrypt(String value) {
        try {
            Cipher cipher = initCipher(Cipher.ENCRYPT_MODE);
            byte[] result = cipher.doFinal(value.getBytes());
            return Base64.encodeToString(result, Base64.DEFAULT);
        } catch (Exception e) {
            e.printStackTrace();
            return "";
        }
    }

    public String decrypt(String value) {
        try {
            Cipher cipher = initCipher(Cipher.DECRYPT_MODE);
            byte[] result = cipher.doFinal(Base64.decode(value, Base64.DEFAULT));
            return new String(result);
        } catch (Exception e) {
            e.printStackTrace();
            return "";
        }
    }

    private Cipher initCipher(int mode) throws Exception {
        Cipher cipher = Cipher.getInstance(getTransformation());
        cipher.init(mode, getKey(), getIvParameterSpec());
        return cipher;
    }

    protected Key getKey(){
        return new SecretKeySpec(getKeyString().getBytes(), getAlgorithm());
    }
}

File: app/src/main/java/com/kailin/utillibrary/security/CipherUtil.kt
package com.kailin.utillibrary.security

import android.util.Base64

import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import java.security.Key

abstract class CipherUtil {

    protected abstract val key: Key?

    protected abstract val ivParameterSpec: IvParameterSpec?

    protected abstract val transformation: String

    protected open val encryptCipher: Cipher
        get() = initCipher(Cipher.ENCRYPT_MODE)

    protected open val decryptCipher: Cipher
        get() = initCipher(Cipher.DECRYPT_MODE)

    fun encrypt(value: String): String {
        return try {
            val result = encryptCipher!!.doFinal(value.toByteArray())
            val base64Encode = Base64.encode(result, Base64.NO_WRAP)
            String(base64Encode)
        } catch (e: Exception) {
            e.printStackTrace()
            ""
        }
    }

    fun decrypt(value: String): String {
        return try {
            val base64Decode = Base64.decode(value, Base64.NO_WRAP)
            val result = decryptCipher!!.doFinal(base64Decode)
            String(result)
        } catch (e: Exception) {
            e.printStackTrace()
            ""
        }
    }

    @Throws(Exception::class)
    protected open fun initCipher(mode: Int): Cipher {
        val cipher = Cipher.getInstance(transformation)
        cipher.init(mode, key, ivParameterSpec)
        return cipher
    }
}
